src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java

```diff
@@ -14,30 +14,158 @@
 @WebServlet("/lecture")
 public class LectureServlet extends HttpServlet {
+  /* Used to generate a Pattern for a Video URL. */
+  private static final String YOUTUBE_VIDEO_URL_PATTERN =
+      "(?<=watch\\?v=|/videos/|embed\\/|youtu.be\\/|\\/v\\/|\\/e\\/|"
+      + "watch\\?v%3D|watch\\?feature=player_embedded&v=|%2Fvideos%2F|"
+      + "embed%\u200C\u200B2F|youtu.be%2F|%2Fv%2F)[^#\\&\\?\\n]*";
+
+  /* Name of input field used for lecture name in lecture selection page. */
+  private static final String NAME_INPUT = "name-input";
+  /* Name of input field used for lecture video link in lecture selection page. */
+  private static final String LINK_INPUT = "link-input";
+  /* Default value if new lecture inputs are empty. */
+  private static final String DEFAULT_VALUE = "";
+  private static final String REDIRECT_URL = "/lecture-view.html";
+
+  /* Pattern used to create a matcher for a video ID. */
+  private static Pattern videoUrlGeneratedPattern;
+  private static DatastoreService datastore;
+
   @Override
   public void init() throws ServletException {
-    // TODO: Implement LectureServlet.
+    datastore = DatastoreServiceFactory.getDatastoreService();
+    videoUrlGeneratedPattern = Pattern.compile(YOUTUBE_VIDEO_URL_PATTERN);
   }
 
   @Override
+  // TODO: Check if URL is valid.
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.sendError(500, "Not Implemented");
+    String videoUrl = getParameter(request, LINK_INPUT, DEFAULT_VALUE);
+    Optional<Entity> existingEntity = checkUrlInDatabase(videoUrl);
+
+    if (existingEntity.isPresent()) {
+      response.sendRedirect(buildRedirectUrl(existingEntity.get()));
+      return;
+    }
+    Entity lectureEntity = createLectureEntity(request);
+    datastore.put(lectureEntity);
+    response.sendRedirect(buildRedirectUrl(lectureEntity));
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.sendError(500, "Not Implemented");
+    List<Lecture> lectures = getLectures();
+    Gson gson = new Gson();
+    response.setContentType("application/json");
+    response.getWriter().println(gson.toJson(lectures));
+  }
+
+  /**
+   * Returns the Entity in database that has {@code url}, or
+   * {@code Optional.empty()} if one doesn't exist.
+   */
+  private Optional<Entity> checkUrlInDatabase(String url) {
+    Query query = new Query(Lecture.ENTITY_KIND);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+    Iterable<Entity> resultsIterable = results.asIterable();
+
+    for (Entity lecture : resultsIterable) {
+      if (lecture.getProperty(Lecture.PROP_URL).equals(url)) {
+        return Optional.of(lecture);
+      }
+    }
+    return Optional.empty();
+  }
+
+  /** Creates and returns {@code lectureEntity} using parameters found in {@code request}. */
```

Comment:
Please do not reference internal code. If someone is reading the JavaDoc, they tend to not have access to the code. Thinks about public API javadocs, they never reference internal code because you cannot read it.

Reply:
fixed